crates/bevy_ui/src/focus.rs

```diff
@@ -56,82 +56,62 @@ pub struct State {
     entities_to_reset: SmallVec<[Entity; 1]>,
 }
 
-/// The system that sets Interaction for all UI elements based on the mouse cursor activity
-#[allow(clippy::type_complexity)]
+pub type NodeQuery<'a> = (
+    Entity,
+    &'a Node,
+    &'a GlobalTransform,
+    &'a mut Interaction,
+    Option<&'a FocusPolicy>,
+    Option<&'a CalculatedClip>,
+);
+
+/// The system that sets Interaction for all UI elements based on the mouse and touch cursor
+/// activity
 pub fn ui_focus_system(
-    mut state: Local<State>,
+    state: Local<State>,
     windows: Res<Windows>,
     mouse_button_input: Res<Input<MouseButton>>,
     touches_input: Res<Touches>,
-    mut node_query: Query<(
-        Entity,
-        &Node,
-        &GlobalTransform,
-        Option<&mut Interaction>,
-        Option<&FocusPolicy>,
-        Option<&CalculatedClip>,
-    )>,
+    node_query: Query<NodeQuery>,
 ) {
-    let cursor_position = windows
-        .get_primary()
-        .and_then(|window| window.cursor_position());
-
-    // reset entities that were both clicked and released in the last frame
-    for entity in state.entities_to_reset.drain(..) {
-        if let Ok(mut interaction) = node_query.get_component_mut::<Interaction>(entity) {
-            *interaction = Interaction::None;
-        }
-    }
+    focus_ui(
+        state,
+        windows,
+        mouse_button_input,
+        touches_input,
+        node_query,
+    )
+}
 
-    let mouse_released =
-        mouse_button_input.just_released(MouseButton::Left) || touches_input.just_released(0);
-    if mouse_released {
-        for (_entity, _node, _global_transform, interaction, _focus_policy, _clip) in
-            node_query.iter_mut()
-        {
-            if let Some(mut interaction) = interaction {
-                if *interaction == Interaction::Clicked {
-                    *interaction = Interaction::None;
-                }
-            }
-        }
-    }
+fn focus_ui<Cursor: CursorResource>(
+    mut state: Local<State>,
+    windows: Res<Cursor>,
+    mouse_button_input: Res<Input<MouseButton>>,
+    touches_input: Res<Touches>,
+    mut node_query: Query<NodeQuery>,
+) {
+    reset_interactions(
```

Comment:
I like this living in a separate system at least.

Reply:
Should I extract it? This system currently does two things:
- resetting Interactions to None
- checking the input to set some Interactions to Hovered / Clicked

The coupling between those two things is quite hard as they rely on each other. This becomes quite obvious from the `Local<State>`. Actually I wanted to get rid of it, but I did not have a good idea without changing the logic.